Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -196,4 +197,139 @@ lemma two_mul_semilength_eq_length : 2 * p.semilength = p.toList.length := by
+  replace f : p.firstReturn = 0 := by omega
+  rw [firstReturn, findIdx_eq (by rw [length_range]; exact length_pos.mpr h)] at f
+  simp only [get_eq_getElem, getElem_range] at f
+  rw [← p.cons_tail_dropLast_concat h] at f
+  simp at f
+
+lemma firstReturn_lt_length (h : p.toList ≠ []) : p.firstReturn < p.toList.length := by
+  have lp := length_pos_of_ne_nil h
+  rw [← length_range p.toList.length]
+  apply findIdx_lt_length_of_exists
+  simp only [mem_range, decide_eq_true_eq]
+  use p.toList.length - 1
+  exact ⟨by omega, by rw [Nat.sub_add_cancel lp, take_of_length_le (le_refl _),
+    p.count_U_eq_count_D]⟩
+
+lemma count_eq_count_of_take_firstReturn_add_one (h : p.toList ≠ []) :
+    (p.toList.take (p.firstReturn + 1)).count U = (p.toList.take (p.firstReturn + 1)).count D := by
+  have := findIdx_get (w := (length_range p.toList.length).symm ▸ p.firstReturn_lt_length h)
+  simpa using this
+
+lemma count_lt_count_of_lt_firstReturn (i : ℕ) (hi : i < p.firstReturn) :
+    (p.toList.take (i + 1)).count D < (p.toList.take (i + 1)).count U := by
+  have ne := not_of_lt_findIdx hi
+  rw [decide_eq_true_eq, ← ne_eq, get_eq_getElem, getElem_range] at ne
+  exact lt_of_le_of_ne (p.count_D_le_count_U (i + 1)) ne.symm
+
+@[simp]
+lemma firstReturn_add : (p + q).firstReturn = if p = 0 then q.firstReturn else p.firstReturn := by
+  split_ifs with h; · simp [h]
+  rw [← toList_eq_nil, ← ne_eq] at h
+  have u : (p + q).toList = p.toList ++ q.toList := rfl
+  have v := p.firstReturn_lt_length h
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_append_eq_append_take (l₂ := q.toList),
+        show p.firstReturn + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil, p.count_eq_count_of_take_firstReturn_add_one h]
+    · intro j hj
+      rw [take_append_eq_append_take, show j + 1 - p.toList.length = 0 by omega,
+        take_zero, append_nil]
+      exact (p.count_lt_count_of_lt_firstReturn j hj).ne'
+  · rw [length_range, u, length_append]
+    exact Nat.lt_add_right _ (p.firstReturn_lt_length h)
+
+@[simp]
+lemma firstReturn_nest : p.nest.firstReturn = p.toList.length + 1 := by
+  have u : p.nest.toList = U :: p.toList ++ [D] := rfl
+  rw [firstReturn, findIdx_eq]
+  · simp_rw [get_eq_getElem, getElem_range, u, decide_eq_true_eq]
+    constructor
+    · rw [take_of_length_le (by simp), ← u, p.nest.count_U_eq_count_D]
+    · intro j hj
+      simp_rw [cons_append, take_cons, count_cons, beq_self_eq_true, ite_true,
+        beq_iff_eq, ite_false, take_append_eq_append_take,
+        show j - p.toList.length = 0 by omega, take_zero, append_nil]
+      have := p.count_D_le_count_U j
+      omega
+  · simp_rw [length_range, u, length_append, length_cons]
+    exact Nat.lt_add_one _
+
+/-- The right part of the Dyck word decomposition. -/
+def rightPart (h : p.toList ≠ []) : DyckWord :=
+  p.drop (p.firstReturn + 1) (p.count_eq_count_of_take_firstReturn_add_one h)
```

Comment:
Can you just return `0` if `p = 0`? In fact, do you think there is a reasonable choice of junk value for `DyckWord.drop` when the subword isn't balanced?

Reply:
I've done the first modification, but I answer no to the second question; in order to preserve `p.take i + p.drop i = p` for all Dyck words and have both summands also Dyck words, if `p` is of the form `q` then one of `p.take i` and `p.drop i` must be empty and the other must be the whole word, and there is no preference for either choice.